### PR TITLE
A0-1455: Refactor for incoming protocol upgrade

### PR DESCRIPTION
**Pull request by timorl**
# Description

Preparatory work for #709, some changes here don't make sense here by themselves, see there for context.

## Type of change

Please delete options that are not relevant.

# Checklist:

### CODE REVIEW

File: finality-aleph/src/validator_network/protocols/v0/mod.rs [hunk 1]
```diff
@@ -217,24 +141,31 @@ mod tests {
...
+        UnboundedReceiver<(
+            AuthorityId,
+            Option<mpsc::UnboundedSender<D>>,
+            ConnectionType,
+        )>,
+        UnboundedReceiver<(
+            AuthorityId,
+            Option<mpsc::UnboundedSender<D>>,
+            ConnectionType,
+        )>,
```
Comment (pmikolajczyk41) [inline finality-aleph/src/validator_network/protocols/v0/mod.rs:153]: just to support my other comment: these should be `UnboundedReceiver<ResultForService<D>>` (or maybe even onshots?)

File: finality-aleph/src/validator_network/manager/mod.rs [hunk 2]
```diff
@@ -48,7 +59,7 @@ impl ManagerStatus {
         let incoming: HashSet<_> = manager
             .incoming
             .iter()
-            .filter(|(_, exit)| !exit.is_canceled())
+            .filter(|(_, exit)| !exit.is_closed())
```
Comment (pmikolajczyk41) [inline finality-aleph/src/validator_network/manager/mod.rs:62]: why this change?

Comment (pmikolajczyk41) [reply finality-aleph/src/validator_network/manager/mod.rs:62]: oh, because of the change from oneshot to unbounded, right? then why this one?

**Review by pmikolajczyk41 — APPROVED** [review]
overall looks good to me, just left a few comments, but feel free to resolve every one of them

I know that many docstrings have already been here, but I have to express my delight - they are great and really helpful

**Review by lesniak43 — APPROVED** [review]

File: finality-aleph/src/validator_network/protocols/mod.rs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+/// Defines the protocol for communication.
+#[derive(Debug, PartialEq, Eq)]
+pub enum Protocol {
+    /// The first version of the protocol, with unidirectional connections.
+    V0,
+}
+
+/// Protocol error.
+#[derive(Debug)]
+pub enum ProtocolError {
```
[the remark this answers is not in the document]
Comment (timorl) [reply finality-aleph/src/validator_network/protocols/mod.rs:50]: This will be the superset of errors. We don't expect to support more than two protocol versions at once, so this should never be too much.

File: finality-aleph/src/validator_network/protocols/v0/mod.rs [hunk 4]
```diff
@@ -91,18 +36,22 @@ async fn sending<D: Data, S: AsyncWrite + Unpin + Send>(
 
 /// Performs the handshake, and then keeps sending data received from the parent service.
 /// Exits on parent request, or in case of broken or dead network connection.
-async fn v0_outgoing<D: Data, S: Splittable>(
+pub async fn outgoing<D: Data, S: Splittable>(
     stream: S,
     authority_pen: AuthorityPen,
     peer_id: AuthorityId,
-    result_for_parent: mpsc::UnboundedSender<(AuthorityId, Option<mpsc::UnboundedSender<D>>)>,
+    result_for_parent: ResultForService<D>,
```
[the remark this answers is not in the document]
Comment (timorl) [reply finality-aleph/src/validator_network/protocols/v0/mod.rs:43]: No, that is one of these things that don't make sense without context. In the incoming PRs we will want to pass a proper unbounded sender to another protocol version, and eliminate the oneshot completely, this is preparation for that.

File: finality-aleph/src/validator_network/protocols/mod.rs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+pub enum ConnectionType {
+    LegacyIncoming,
+    LegacyOutgoing,
+}
+
+/// What connections send back to the service after they become established. Starts with a peer id
+/// of the remote node, followed by a channel for sending data to that node, with None if the
+/// connection was unsuccessful and should be reestablished. Finally a marker for legacy
+/// compatibility.
+pub type ResultForService<D> = mpsc::UnboundedSender<(
```
[the remark this answers is not in the document]
Comment (timorl) [reply finality-aleph/src/validator_network/protocols/mod.rs:35]: Hm, right, makes sense. Fixed.

File: finality-aleph/src/validator_network/protocols/v0/mod.rs (hunk 1 above)
Comment (timorl) [reply finality-aleph/src/validator_network/protocols/v0/mod.rs]: Partially fixed, partially see comment above.

File: finality-aleph/src/validator_network/manager/mod.rs (hunk 2 above)
Comment (timorl) [reply finality-aleph/src/validator_network/manager/mod.rs:62]: See comment above.
